File: services/ai-fastapi/app/core/evaluator.py
"""Assessment evaluator integrating LLM grading, BKT mastery, and sentiment analysis.

Pipeline for each submission
-----------------------------
1. Fetch the node's ``assessment_rubric`` from ``GraphEngine``.
2. Run sentiment analysis on the student's answer (D2).
3. Build LLM prompts — tone-adjusted if frustration is detected (D2).
4. Grade the answer with the LLM (existing flow).
5. Convert the LLM score to a binary signal and update BKT mastery (D1).
6. Return a rich response that includes score, BKT mastery probability,
   sentiment, and Socratic feedback.

Architecture boundary
---------------------
This service *proposes* mastery (``mastered: bool``, ``p_mastery: float``).
The Go orchestration service remains the final authority for writing the
official Competency Record and unlocking graph nodes.
"""

from typing import Any

from app.core.bkt_estimator import BKTEstimator, get_bkt_estimator
from app.core.graph_engine import GraphEngine
from app.core.llm_client import LLMClient
from app.core.sentiment_analyzer import (
    SentimentAnalyzer,
    build_tone_system_prompt,
    get_sentiment_analyzer,
)

# Score threshold for converting a graded score into a BKT binary signal.
# A score >= this value is treated as a "correct" response (1), else "incorrect" (0).
SCORE_CORRECT_THRESHOLD: float = 0.7


class AssessmentEvaluator:
    """Evaluate learner submissions using LLM + BKT + sentiment analysis.

    The evaluator fetches the node's ``assessment_rubric`` from ``GraphEngine``,
    prompts the LLM (with sentiment-adapted tone) to score the answer, converts
    the score into a binary BKT signal, and returns a normalized JSON result
    enriched with mastery probability and emotional context.
    """

    def __init__(
        self,
        llm_client: LLMClient,
        graph_engine: GraphEngine,
        bkt_estimator: BKTEstimator | None = None,
        sentiment_analyzer: SentimentAnalyzer | None = None,
    ) -> None:
        self.llm = llm_client
        self.graph = graph_engine
        self.bkt = bkt_estimator or get_bkt_estimator()
        self.sentiment = sentiment_analyzer or get_sentiment_analyzer()

    def evaluate_submission(
        self,
        domain_id: str,
        node_id: str,
        student_answer: str,
        attempt_history: list[int] | None = None,
    ) -> dict[str, Any]:
        """Evaluate a student's answer for a specific domain node.

        Args:
            domain_id:       The domain identifier containing the node.
            node_id:         The node identifier that carries the assessment rubric.
            student_answer:  The learner's submitted answer (text).
            attempt_history: Optional list of previous binary responses for this
                             node (1=correct, 0=incorrect).  When provided, BKT
                             mastery is computed across the full history + this
                             attempt.  When omitted, only this attempt is used.

        Returns:
            A dict containing:
                - ``score``           — normalized float 0.0–1.0
                - ``passed``          — bool (score >= SCORE_CORRECT_THRESHOLD)
                - ``feedback``        — Socratic feedback string
                - ``remediation_hint``— short actionable hint (only if failed)
                - ``bkt``             — BKT mastery result dict (p_mastery, mastered, …)
                - ``sentiment``       — emotion analysis dict
                - ``binary_signal``   — the 0/1 signal fed into BKT
        """

        # ------------------------------------------------------------------ #
        # 1. Retrieve node and rubric
        # ------------------------------------------------------------------ #
        try:
            node = self.graph._node_lookup[domain_id][node_id]
        except KeyError as exc:
            raise ValueError(f"Node '{node_id}' not found in domain '{domain_id}'.") from exc

        rubric = node.get("assessment_rubric") or {}
        diagnostic_question = rubric.get("diagnostic_question", "Please evaluate the following response.")
        key_criteria = rubric.get("key_evaluation_criteria", [])
        remediation_hint_default = rubric.get("remediation_hint", "Consider reviewing the core concepts and examples.")

        # Pull any per-node BKT params from the graph and register them
        bkt_params = node.get("bkt_params")
        if bkt_params and isinstance(bkt_params, dict):
            self.bkt.register_skill(
                node_id,
                p_init=bkt_params.get("p_init", 0.30),
                p_learn=bkt_params.get("p_learn", 0.20),
                p_guess=bkt_params.get("p_guess", 0.25),
                p_slip=bkt_params.get("p_slip", 0.10),
            )

        # ------------------------------------------------------------------ #
        # 2. Sentiment analysis (D2) — run BEFORE prompting the LLM
        # ------------------------------------------------------------------ #
        sentiment_result = self.sentiment.analyze(student_answer)
        tone_override = sentiment_result.get("tone_override", "standard")

        # ------------------------------------------------------------------ #
        # 3. Build LLM prompts — tone-adjusted if frustration is detected
        # ------------------------------------------------------------------ #
        base_system_prompt = (
            "You are a strict pedagogical evaluator. Grade answers against the provided criteria."
            " Provide a numeric score between 0.0 and 1.0, concise constructive feedback, and a remediation hint if the"
            " answer does not meet the threshold. Be objective and base the score on the criteria only."
        )
        system_prompt = build_tone_system_prompt(base_system_prompt, tone_override)

        criteria_text = (
            "\n".join([f"{i+1}. {c}" for i, c in enumerate(key_criteria)])
            if key_criteria
            else "(no explicit criteria provided)"
        )

        user_prompt = (
            f"Diagnostic question: {diagnostic_question}\n"
            f"Key evaluation criteria:\n{criteria_text}\n\n"
            f"Student answer:\n{student_answer}\n\n"
            "Return ONLY valid JSON with the following fields: \n"
            "- score: a number between 0.0 and 1.0\n"
            f"- passed: boolean (true if score >= {SCORE_CORRECT_THRESHOLD})\n"
            "- feedback: constructive, Socratic feedback guiding the student\n"
            "- remediation_hint: (optional) short actionable hint if failed\n"
        )

        response_schema = {
            "score": {"type": "number", "min": 0.0, "max": 1.0},
            "passed": {"type": "boolean"},
            "feedback": {"type": "string"},
            "remediation_hint": {"type": "string"},
        }

        # ------------------------------------------------------------------ #
        # 4. LLM grading
        # ------------------------------------------------------------------ #
        result = self.llm.generate_structured_json(
            system_prompt=system_prompt,
            user_prompt=user_prompt,
            response_schema=response_schema,
        )

        if isinstance(result, dict) and result.get("error"):
            return {"error": result.get("error"), "raw": result.get("raw")}

        # Normalize score
        try:
            score_val = float(result.get("score", 0.0))
        except (TypeError, ValueError):
            return {"error": "Invalid or missing 'score' in LLM response.", "raw": result}

        score = max(0.0, min(1.0, score_val))
        passed = score >= SCORE_CORRECT_THRESHOLD

        feedback = result.get("feedback") or "No feedback provided."
        remediation = result.get("remediation_hint") or (remediation_hint_default if not passed else None)

        # ------------------------------------------------------------------ #
        # 5. BKT mastery update (D1)
        # ------------------------------------------------------------------ #
        binary_signal = 1 if passed else 0

        # Build full history: previous attempts + this one
        history = list(attempt_history) if attempt_history else []
        history.append(binary_signal)

        bkt_result = self.bkt.estimate(node_id=node_id, attempt_history=history)

        # ------------------------------------------------------------------ #
        # 6. Assemble and return the enriched response
        # ------------------------------------------------------------------ #
        response: dict[str, Any] = {
            "score": round(score, 3),
            "passed": passed,
            "feedback": feedback,
            "binary_signal": binary_signal,
            "bkt": bkt_result,
            "sentiment": sentiment_result,
        }

        if not passed and remediation:
            response["remediation_hint"] = remediation

        return response

File: services/api-go/internal/aiengine/evaluator.go
package aiengine

// AssessmentEvaluator is a faithful Go port of app/core/evaluator.py.

import (
	"fmt"
	"math"
	"strings"
)

// SCORE_CORRECT_THRESHOLD matches the Python evaluator.
const scoreCorrectThreshold = 0.7

const evalBaseSystemPrompt = "" +
	"You are a strict pedagogical evaluator. Grade answers against the provided criteria." +
	" Provide a numeric score between 0.0 and 1.0, concise constructive feedback, and a remediation hint if the" +
	" answer does not meet the threshold. Be objective and base the score on the criteria only."

// EvaluateSubmission mirrors AssessmentEvaluator.evaluate_submission().
func (engine *GraphEngine) EvaluateSubmission(domainID, nodeID, studentAnswer string, attemptHistory []int, llm *LLMClient) map[string]any {
	// 1. Retrieve node and rubric.
	node, err := engine.GetNode(domainID, nodeID)
	if err != nil {
		return map[string]any{"error": err.Error()}
	}
	rubric := node.AssessmentRubric
	diagnosticQuestion, _ := rubric["diagnostic_question"].(string)
	if diagnosticQuestion == "" {
		diagnosticQuestion = "Please evaluate the following response."
	}
	keyCriteria := anyStringSlice(rubric["key_evaluation_criteria"])
	remediationHintDefault, _ := rubric["remediation_hint"].(string)
	if remediationHintDefault == "" {
		remediationHintDefault = "Consider reviewing the core concepts and examples."
	}

	// Pull per-node BKT params from the graph and register them.
	if bpRaw, ok := rubric["bkt_params"].(map[string]any); ok {
		bp := DefaultBktParams()
		if v, ok := num(bpRaw["p_init"]); ok {
			bp.PInit = v
		}
		if v, ok := num(bpRaw["p_learn"]); ok {
			bp.PLearn = v
		}
		if v, ok := num(bpRaw["p_guess"]); ok {
			bp.PGuess = v
		}
		if v, ok := num(bpRaw["p_slip"]); ok {
			bp.PSlip = v
		}
		RegisterSkill(nodeID, bp)
	}

	// 2. Sentiment analysis BEFORE prompting the LLM.
	sentiment := AnalyzeSentiment(studentAnswer)
	toneOverride := sentiment.ToneOverride

	// 3. Build LLM prompts — tone-adjusted if frustration detected.
	systemPrompt := BuildToneSystemPrompt(evalBaseSystemPrompt, toneOverride)

	var criteriaText string
	if len(keyCriteria) > 0 {
		var b strings.Builder
		for i, c := range keyCriteria {
			fmt.Fprintf(&b, "%d. %s\n", i+1, c)
		}
		criteriaText = b.String()
	} else {
		criteriaText = "(no explicit criteria provided)"
	}

	userPrompt := fmt.Sprintf(
		"Diagnostic question: %s\nKey evaluation criteria:\n%s\n\nStudent answer:\n%s\n\n"+
			"Return ONLY valid JSON with the following fields: \n"+
			"- score: a number between 0.0 and 1.0\n"+
			"- passed: boolean (true if score >= %v)\n"+
			"- feedback: constructive, Socratic feedback guiding the student\n"+
			"- remediation_hint: (optional) short actionable hint if failed\n",
		diagnosticQuestion, criteriaText, studentAnswer, scoreCorrectThreshold,
	)

	schema := map[string]any{
		"score":            map[string]any{"type": "number", "min": 0.0, "max": 1.0},
		"passed":           map[string]any{"type": "boolean"},
		"feedback":         map[string]any{"type": "string"},
		"remediation_hint": map[string]any{"type": "string"},
	}

	// 4. LLM grading.
	result := llm.GenerateStructuredJSON(systemPrompt, userPrompt, schema)
	if e, ok := result["error"]; ok && e != nil {
		return map[string]any{"error": e, "raw": result["raw"]}
	}

	// Normalize score.
	scoreVal, ok := num(result["score"])
	if !ok {
		return map[string]any{"error": "Invalid or missing 'score' in LLM response.", "raw": result}
	}
	score := math.Max(0.0, math.Min(1.0, scoreVal))
	passed := score >= scoreCorrectThreshold

	feedback, _ := result["feedback"].(string)
	if feedback == "" {
		feedback = "No feedback provided."
	}
	remediation, _ := result["remediation_hint"].(string)
	if remediation == "" && !passed {
		remediation = remediationHintDefault
	}

	// 5. BKT mastery update.
	binarySignal := 0
	if passed {
		binarySignal = 1
	}
	history := append([]int(nil), attemptHistory...)
	history = append(history, binarySignal)
	bktResult := BktEstimate(nodeID, history, nil)

	// 6. Assemble enriched response.
	response := map[string]any{
		"score":         round3(score),
		"passed":        passed,
		"feedback":      feedback,
		"binary_signal": binarySignal,
		"bkt":           bktResult,
		"sentiment":     sentiment,
	}
	if !passed && remediation != "" {
		response["remediation_hint"] = remediation
	}
	return response
}

func round3(v float64) float64 {
	return math.Round(v*1000) / 1000
}

// tryEvaluateBetter handles the case where the LLM is unavailable and returns
// a deterministic rubric-based score, keeping the evaluate endpoint useful.
// Note: this is intentionally NOT used on the primary path so that BKT and
// mastery semantics stay identical to FastAPI.

func num(v any) (float64, bool) {
	switch t := v.(type) {
	case float64:
		return t, true
	case float32:
		return float64(t), true
	case int:
		return float64(t), true
	case int64:
		return float64(t), true
	case string:
		var f float64
		_, err := fmt.Sscanf(t, "%f", &f)
		return f, err == nil
	}
	return 0, false
}

func anyStringSlice(v any) []string {
	arr, ok := v.([]any)
	if !ok {
		return nil
	}
	var out []string
	for _, item := range arr {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}
